homeassistant/components/icloud/account.py

```diff
@@ -110,9 +115,36 @@ def setup(self) -> None:
                 self._icloud_dir.path,
                 with_family=self._with_family,
             )
-        except PyiCloudFailedLoginException as error:
+        except PyiCloudFailedLoginException:
             self.api = None
-            _LOGGER.error("Error logging into iCloud Service: %s", error)
+            # Login failed which means credentials need to be updated.
+            username = self._config_entry.data[CONF_USERNAME]
+            message = (
+                f"Your password for '{username}' is no longer working. Go to the "
+                "Integrations menu and click on Configure on the discovered Apple "
+                "iCloud card to login again."
+            )
+            _LOGGER.error(message)
+            self.hass.components.persistent_notification.async_create(
+                message,
+                asyncio.run_coroutine_threadsafe(
+                    async_get_integration(self.hass, DOMAIN), self.hass.loop
+                )
+                .result()
+                .manifest["name"],
+                f"{self._config_entry.entry_id}_{SOURCE_REAUTH}",
+            )
+
+            self.hass.async_create_task(
```

Comment:
You should just put all the handling of reauth needed in an async function instead of calling into async context multiple times

Reply:
Since I removed creating the notification, I don't think this still applies

Reply:
We're in sync context here. We can not use the async api directly. So the comment still applies.

Reply:
Is the issue that `async_create_task` can't be called in a sync context? Do I need to use `asyncio.run_coroutine_threadsafe`? I originally interpreted this comment as "you have multiple async calls in a sync context - put them all into a single async function so you are only making a single async call from the sync context"

Reply:
This should work.
```suggestion
            self.hass.add_job(
```

Reply:
thank you!